src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.cs

```diff
@@ -141,9 +141,10 @@ RaspberryBoardInfo.Model.RaspberryPi3BPlus or
             RaspberryBoardInfo.Model.RaspberryPiZeroW or
             RaspberryBoardInfo.Model.RaspberryPiZero2W or
             RaspberryBoardInfo.Model.RaspberryPi4 or
+            RaspberryBoardInfo.Model.RaspberryPi5 or
             RaspberryBoardInfo.Model.RaspberryPi400 => new RaspberryPi3LinuxDriver(),

```

Comment:
The GPIO Hardware of the RPi5 seems to be quite different from previous versions. Is the RaspberryPi3LinuxDriver really working correctly?

Reply:
Fair point... I'll see if I can build it all locally and try it on my Pi5 and see what's what. :-)

Reply:
I read the pre-release datasheet for the new custom IC called `RP1` which is new to the RPi5.
It governs all the I/O, serial ports, I2C, SPI, PWM, video and USB communications. The HW architecture is very different from the previous boards even if they apparently maintained the external pin compatibility (They also use MUXes for the I/O, meaning that you can remap the ping assignments).

The main point about that IC is to offload the communications from the main processor to support low-power communication while the rest of the board is in sleep mode. It also supports 1.8V levels but they have to be carefully used because, once set any voltage higher than 1.8V can damage the board.

Said that, the libgpiod could be enough to decouple the new functionalities but since there are more features, it will probably deserve a new separate `RasperryPi5LinuxDriver` to support them.

I don't have a new RPi5, but any test you could do is more than welcome :)

Reply:
Is there a recommended workflow for development / debugging for GPIO? Are we able to debug directly on the Pi? @pgrawehr @raffaeler ...

Reply:
> Is there a recommended workflow for development / debugging for GPIO? Are we able to debug directly on the Pi? 

The point here is that there is no direct access for the C# code to the GPIO, therefore debugging the GPIO access does not help much.  The 'dotnet/iot` library provides two "drivers": `sysfs` (GPIOs mapped on files) and `libgpiod`. They are just the C# code accessing the file system or the `libgpiod` api.

- `sysfs` is the simpler (and slower) way. If you can successfully access the GPIOs using the default tools, it should work from `dotnet/iot` as well. Even if this works, it is not the best solution as file mapped GPIOs have a number of issues.
- `libgpiod` provides access to the GPIOs via APIs. I understood that there is no current version for `libgpiod` for the RPi5. In this case there is nothing you can do from dotnet. A soon as the Raspberry Foundation should provide it, we will have to see if the APIs are exactly the same or not.

Alternate solutions are:
- fork `libgpiod` and provide support for the `RPi5`, you should download and implement the `RP1` specifications inside the `libgpiod` source code. This step is definitely not trivial.
- implement the `RP1` specs directly in C# bypassing libgpiod. It is doable but, again, not trivial.
AFAIK we never did this for any other board.

P.S. As I mentioned in the previous message the RP1 specs have a number of new features that may not fit in the current `libgpiod` API. This is why I believe we need something different on the C# wrapper side as well.

Reply:
Gotcha... Not trivial is definitely a takeaway here, especially given that I've not been involved in the code at all here.

Having a quick play...

## `sysfs`

 I can see that `sysfs` looks like it's able to access the GPIO... 

![image](https://github.com/dotnet/iot/assets/7139618/7fbfc4be-6e55-466f-8457-abf0e4276b3e)

However, it doesn't actually manage to change the physical state of pins... Of course, this is deprecated in favour of `libgpiod`...

## `pinctrl`

`pinctrl` seems to work fine;

![image](https://github.com/dotnet/iot/assets/7139618/aef56f98-b5b0-4f80-95cd-3ac8ef51bc63)

With the pin happily setting and clearing.

## `gpiod`

Installing gpiod and running gpiodetect I get;

![image](https://github.com/dotnet/iot/assets/7139618/de0b929b-9ea1-4923-b005-82b3e7a80ced)

gpioinfo for pinctrl-rp1 gives;

![image](https://github.com/dotnet/iot/assets/7139618/6a8897ce-8b15-49e5-837b-12b3c5ee7b65)

And I'm able to toggle GPIO15 with gpioset;

![image](https://github.com/dotnet/iot/assets/7139618/f317daa1-57d8-4627-a39e-d80c110b564b)

## Python `GPIOZero`

As a side note... Python GPIOZero is functioning correctly.

(I think there's a blob post waiting to happen here!)

Reply:
Great work!
From your tests, it looks like `libgpiod` is already working, which is great.
I also suspect that `sysfs` writes did not work for some other reasons. Once the GPIO is mapped, read or write should no make any difference, unless you don't have the appropriate permissions. Try as `root`.

At this point you can make the test you initially planned. Try creating the `GpioController` explicitly passing the `LibGpiodDriver` instance and see if you can read/write the GPIOs.

Reply:
Cool beans and thanks!

This side of things is all quite new to me to be fair, not played with GPIO at this level before., very interesting indeed.

I'll have a play and see what I can do!

I'm guessing I'll have to clone, switch to my working branch, and build this on the pi, then manually add the system.device.gpio dlls to my project to begin with...?

Reply:
@pjgpetecodes There are two options to work with the library:
##Variant 1
- Check out the repo on the PC, build
- Create / update an example
- Publish it for arm64 and copy to the pi
- Run on the Pi, use SSH for remote debugging, if needed

##Variant 2
- Checkout on the PI
- Build on the Pi
- Execute the examples / Integration tests

I'm normally going with 1, unless I'm running the GPIO testbench. Then I check out on the Pi and run `dotnet test System.Device.Gpio.sln`. (For the GPIO tests to work, you need to have pins 5 and 6 connected with a resistor)

Reply:
> I'm guessing I'll have to clone, switch to my working branch, and build this on the pi, then manually add the system.device.gpio dlls to my project to begin with...?

You can just cross-compiling everything on Windows and then copy to the RPi5. Your test code could be any basic binding. I suggest you to directly reference the dotnet/iot libraries so that you recompile the libraries as well. 

This is the typical way to cross-compile using self-contained to avoid any dependency and being able to sync the sources as suggested by Patrick.
- Release vs Debug: pick Debug if you want to avoid optimizations and sync the sources
- linx-arm vs linux-arm64. This depends on the OS on the RPi

```
dotnet deploy -c Release -r linux-arm --self-contained=true
``` 
Please note that the pdbs will point to sources that you just compiled.

In any case, if you just want to make a test, you don't need to debug. You can just deploy the release version as self-contained and see if if works (you can write on the console some message to have detailed info).

HTH

Reply:
Arghhh... My comments above were stuck in pending... Apologies @raffaeler and @pgrawehr ... Thanks for the help mind you!

Reply:
Hey, that's good news, at least we have one driver that works for the Pi5. Can you (just for reference) do the same test with using the RaspberryPi3Driver and the SysFsDriver?

PinNumberingScheme.Board is only supported with the RaspberryPi3Driver, because it needs to know the mapping from the logical pins to the physical pins. Because most boards and drivers don't support that either, we've mostly given up support for that anyway and stick to the logical numbering.

Reply:
So... 

With the original code;

- `RaspberryPi3Driver` driver gives `Not a supported Raspberry Pi type: ` (Fairly obviously)`
- `SysFSDriver` just works!

With Raspberry Pi 5 added in the `boardInfo.BoardModel` switch in the `CreateInternalRaspberryPi3LinuxDriver` function, creating the `RaspberryPi3Driver` and passing it into the GpioController gives `Error 13 Initializing the Gpio driver`